### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [174]:
# Dependencies and Setup
import pandas as pd
import os
import numpy as np

In [175]:
# import file
path = os.path.join("Resources","purchase_data.csv")
df = pd.read_csv(path)

## Player Count

* Display the total number of players


In [176]:
player_count = len(df["SN"].unique())
print("Unique player count: {}".format(player_count))

Unique player count: 576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [177]:
# purchasing analysis

# number of unique items bought
unique_purchases = len(df["Item ID"].unique())

# avg purchase price - avg across all items
avg_pur_price = df["Price"].mean()

# total number of purchases.
# method: each row is a purchase, just get the number of rows
total_pur = df.shape[0]

# total revenue
total_rev = df["Price"].sum()

# print nicely
cols = ["Number of Unique Items","Average Price","Number of Purchases","Total Revenue"]

df2 = pd.DataFrame(data=[unique_purchases,avg_pur_price,total_pur,total_rev]).transpose()
df2.columns = cols
df2

Number of Unique Items  Average Price  Number of Purchases  Total Revenue
0                   183.0       3.050987                780.0        2379.77

## Gender Demographics

In [178]:
df2 = df[["Gender","SN"]]
df2.drop_duplicates(keep="first",inplace=True)
df3 = df2.groupby("Gender")

counts = df3["Gender"].count()
pct = counts / sum(counts)

df4 = pd.concat([counts,pct],axis = 1)
df4.columns = ["Total count","Percent of players"]
df4["Percent of players"] = pd.Series(["{0:.2f}%".format(val * 100) for val in df4['Percent of players']], index = df4.index)
df4

c:\anaconda3\lib\site-packages\pandas\util\decorators.py:91: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return func(*args, **kwargs)


Total count Percent of players
Gender                                               
Female                          81             14.06%
Male                           484             84.03%
Other / Non-Disclosed           11              1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [179]:
df2 = df.groupby("Gender")
purchase_count = df2["SN"].count()

avg_purch_price = df2["Price"].mean()

total_purch = df2["Price"].sum()


df3 = df.groupby(["Gender","SN"])
df4 = df3["Price"].sum().to_frame()
df4["Gender"] = [i[0] for i in df4.index]
avg_total_pp = df4.groupby("Gender")["Price"].mean()

df5 = pd.concat([purchase_count,avg_purch_price,total_purch,avg_total_pp],axis = 1)
df5.columns = ["Purchase Count", "Average Purchase Price", "Total Purchase Value","Avg Total Purchase per Person"]
df5

Purchase Count  Average Purchase Price  \
Gender                                                          
Female                            113                3.203009   
Male                              652                3.017853   
Other / Non-Disclosed              15                3.346000   

                       Total Purchase Value  Avg Total Purchase per Person  
Gender                                                                      
Female                               361.94                       4.468395  
Male                                1967.64                       4.065372  
Other / Non-Disclosed                 50.19                       4.562727

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [180]:
df2 = df[["Gender","SN","Age"]]
df2.drop_duplicates(keep="first",inplace=True)

bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
labels = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]
df2["Age_brackets"] = pd.cut(df2["Age"],bins=bins,labels=labels)

df3 = df2.groupby("Age_brackets")

counts = df3["Age_brackets"].count()
pct = counts / sum(counts)

df4 = pd.concat([counts,pct],axis = 1)
df4.columns = ["Total count","Percent of players"]
df4

c:\anaconda3\lib\site-packages\pandas\util\decorators.py:91: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return func(*args, **kwargs)
c:\anaconda3\lib\site-packages\ipykernel\__main__.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Total count  Percent of players
Age_brackets                                 
<10                    17            0.029514
10-14                  22            0.038194
15-19                 107            0.185764
20-24                 258            0.447917
25-29                  77            0.133681
30-34                  52            0.090278
35-39                  31            0.053819
40+                    12            0.020833

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [181]:
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
labels = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

df2 = df.copy()
df2["Age_brackets"] = pd.cut(df2["Age"],bins=bins,labels=labels)

df3 = df2.groupby("Age_brackets")

purch_count = df3["Age_brackets"].count()
avg_price = df3["Price"].mean()
tot_price = df3["Price"].sum()

df4 = df2.groupby(["Age_brackets","SN"])
df5 = df4["Price"].sum().to_frame()
df5["Age_brackets"] = [i[0] for i in df5.index]
avg_tot_price = df5.groupby("Age_brackets")["Price"].mean()

df6 = pd.concat([purch_count,avg_price,tot_price,avg_tot_price],axis=1)
df6.columns = ["Purchase Count","Average Purchase Price","Total Purchase Value","Avg Total Purchase per Person"]
df6.loc[labels]


Purchase Count  Average Purchase Price  Total Purchase Value  \
Age_brackets                                                                 
<10                       23                3.353478                 77.13   
10-14                     28                2.956429                 82.78   
15-19                    136                3.035956                412.89   
20-24                    365                3.052219               1114.06   
25-29                    101                2.900990                293.00   
30-34                     73                2.931507                214.00   
35-39                     41                3.601707                147.67   
40+                       13                2.941538                 38.24   

              Avg Total Purchase per Person  
Age_brackets                                 
<10                                4.537059  
10-14                              3.762727  
15-19                              3.858785  
20-24                              4.318062  
25-29                              3.805195  
30-34                              4.115385  
35-39                              4.763548  
40+                                3.186667

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [182]:
df2 = df.groupby("SN")
pur_count = df2["SN"].count()
avg_purch_price = df2["Price"].mean()

df3 = pd.concat([pur_count,avg_purch_price],axis = 1)
df3.columns = ["Purchase Count","Average Purchase Price"]
df3["Total purchase value"] = df3["Purchase Count"] * df3["Average Purchase Price"]

df3.sort(columns = "Total purchase value",ascending = False).head()

c:\anaconda3\lib\site-packages\ipykernel\__main__.py:9: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


Purchase Count  Average Purchase Price  Total purchase value
SN                                                                       
Lisosia93                 5                3.792000                 18.96
Idastidru52               4                3.862500                 15.45
Chamjask73                3                4.610000                 13.83
Iral74                    4                3.405000                 13.62
Iskadarya95               3                4.366667                 13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [183]:
df2 = df.groupby(["Item Name","Item ID"])
purch = df2["Item Name"].count()
price = df2["Price"].mean()

df3 = pd.concat([purch,price],axis = 1)
df3.columns = ["Purchase Count","Item Price"]
df3 = df3.sort(["Purchase Count"],ascending = False)
df3["TotalPurchaseValue"] = df3["Purchase Count"] * df3["Item Price"]
df3.head()

c:\anaconda3\lib\site-packages\ipykernel\__main__.py:7: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


Purchase Count  \
Item Name                                    Item ID                   
Oathbreaker, Last Hope of the Breaking Storm 178                  12   
Extraction, Quickblade Of Trembling Hands    108                   9   
Nirvana                                      82                    9   
Fiery Glass Crusader                         145                   9   
Pursuit, Cudgel of Necromancy                19                    8   

                                                      Item Price  \
Item Name                                    Item ID               
Oathbreaker, Last Hope of the Breaking Storm 178            4.23   
Extraction, Quickblade Of Trembling Hands    108            3.53   
Nirvana                                      82             4.90   
Fiery Glass Crusader                         145            4.58   
Pursuit, Cudgel of Necromancy                19             1.02   

                                                      TotalPurchaseValue  
Item Name                                    Item ID                      
Oathbreaker, Last Hope of the Breaking Storm 178                   50.76  
Extraction, Quickblade Of Trembling Hands    108                   31.77  
Nirvana                                      82                    44.10  
Fiery Glass Crusader                         145                   41.22  
Pursuit, Cudgel of Necromancy                19                     8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [184]:
df3.sort(["TotalPurchaseValue"],ascending = False).head()

c:\anaconda3\lib\site-packages\ipykernel\__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


Purchase Count  \
Item Name                                    Item ID                   
Oathbreaker, Last Hope of the Breaking Storm 178                  12   
Nirvana                                      82                    9   
Fiery Glass Crusader                         145                   9   
Final Critic                                 92                    8   
Singed Scalpel                               103                   8   

                                                      Item Price  \
Item Name                                    Item ID               
Oathbreaker, Last Hope of the Breaking Storm 178            4.23   
Nirvana                                      82             4.90   
Fiery Glass Crusader                         145            4.58   
Final Critic                                 92             4.88   
Singed Scalpel                               103            4.35   

                                                      TotalPurchaseValue  
Item Name                                    Item ID                      
Oathbreaker, Last Hope of the Breaking Storm 178                   50.76  
Nirvana                                      82                    44.10  
Fiery Glass Crusader                         145                   41.22  
Final Critic                                 92                    39.04  
Singed Scalpel                               103                   34.80

Observations

1. 84% of the purchases are made by males. this is not surprising as the gaming world is dominated by males
2. Avg purchases by a single player is 780/576 = 1.35 -- This number is not very high. Plus we don't have any idea about the number of players who did not buy any items
3. The age group 15-25 represent about 62% of the total purchases. seems like young people play this game more. This could be a trend in the gaming industry in general